pkg/sriov/networkservice/common/filtermechanisms/server.go

```
@@ -0,0 +1,89 @@
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/vfio"
+
+	"github.com/networkservicemesh/sdk-sriov/pkg/sriov"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type filterMechanismsServer struct {
+	config   *sriov.Config
+	mechUsed *sync.Map
+}
+
+// NewServer - filters out mechanisms by type and provided pci address parametr
+func NewServer(config *sriov.Config, mechUsed *sync.Map) networkservice.NetworkServiceServer {
+	return &filterMechanismsServer{config: config, mechUsed: mechUsed}
+}
+
+// Search pci device in config
+// TODO add hostName, capability as parameters for filtering
+func isSupportedPci(pciAddress string, domains []sriov.ResourceDomain) bool {
+	for _, domain := range domains {
+		for _, pciDevice := range domain.PCIDevices {
+			if pciDevice.PCIAddress == pciAddress {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (f *filterMechanismsServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	if request.GetConnection() == nil {
+		request.Connection = &networkservice.Connection{Id: "Id"}
+	}
+	// check on contains
+	if mech, ok := f.mechUsed.Load(request.GetConnection().GetId()); ok {
+		request.Connection.Mechanism = mech.(*networkservice.Mechanism)
+		return request.GetConnection(), nil
+	}
+
+	var mechanisms []*networkservice.Mechanism
+	for _, mechanism := range request.GetMechanismPreferences() {
+		pciAddress := ""
+		switch mechanism.GetType() {
+		case kernel.MECHANISM:
+			pciAddress = kernel.ToMechanism(mechanism).GetPCIAddress()
+		case vfio.MECHANISM:
+			pciAddress = vfio.ToMechanism(mechanism).GetPCIAddress()
+		}
+
+		if pciAddress != "" && isSupportedPci(pciAddress, f.config.Domains) {
+			mechanisms = append(mechanisms, mechanism)
+		}
+	}
+	request.MechanismPreferences = mechanisms
```

Do we need to return an error if len(mechanisms)==0?